File: frontend/src/lib/budgetApi.js
import { RequestJson, RequestWithAuth } from "./apiClient.js";

export const FetchIncomeStreams = async () => {
  return RequestJson("/budget/income-streams");
};

export const CreateIncomeStream = async (payload) => {
  return RequestJson("/budget/income-streams", {
    method: "POST",
    body: JSON.stringify(payload)
  });
};

export const UpdateIncomeStream = async (streamId, payload) => {
  return RequestJson(`/budget/income-streams/${streamId}`, {
    method: "PUT",
    body: JSON.stringify(payload)
  });
};

export const DeleteIncomeStream = async (streamId) => {
  const response = await RequestWithAuth(`/budget/income-streams/${streamId}`, { method: "DELETE" });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || "Request failed");
  }
};

export const FetchExpenses = async () => {
  return RequestJson("/budget/expenses");
};

export const FetchAllocationAccounts = async () => {
  return RequestJson("/budget/allocation-accounts");
};

export const CreateAllocationAccount = async (payload) => {
  return RequestJson("/budget/allocation-accounts", {
    method: "POST",
    body: JSON.stringify(payload)
  });
};

export const UpdateAllocationAccount = async (accountId, payload) => {
  return RequestJson(`/budget/allocation-accounts/${accountId}`, {
    method: "PUT",
    body: JSON.stringify(payload)
  });
};

export const DeleteAllocationAccount = async (accountId) => {
  const response = await RequestWithAuth(`/budget/allocation-accounts/${accountId}`, {
    method: "DELETE"
  });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || "Request failed");
  }
};

export const CreateExpense = async (payload) => {
  return RequestJson("/budget/expenses", {
    method: "POST",
    body: JSON.stringify(payload)
  });
};

export const UpdateExpense = async (expenseId, payload) => {
  return RequestJson(`/budget/expenses/${expenseId}`, {
    method: "PUT",
    body: JSON.stringify(payload)
  });
};

export const UpdateExpenseOrder = async (payload) => {
  const response = await RequestWithAuth("/budget/expenses/order", {
    method: "PUT",
    body: JSON.stringify(payload)
  });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || "Request failed");
  }
};

export const DeleteExpense = async (expenseId) => {
  const response = await RequestWithAuth(`/budget/expenses/${expenseId}`, { method: "DELETE" });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || "Request failed");
  }
};

export const FetchExpenseAccounts = async () => {
  return RequestJson("/budget/expense-accounts");
};

export const CreateExpenseAccount = async (payload) => {
  return RequestJson("/budget/expense-accounts", {
    method: "POST",
    body: JSON.stringify(payload)
  });
};

export const UpdateExpenseAccount = async (accountId, payload) => {
  return RequestJson(`/budget/expense-accounts/${accountId}`, {
    method: "PUT",
    body: JSON.stringify(payload)
  });
};

export const DeleteExpenseAccount = async (accountId) => {
  const response = await RequestWithAuth(`/budget/expense-accounts/${accountId}`, { method: "DELETE" });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || "Request failed");
  }
};

export const FetchExpenseTypes = async () => {
  return RequestJson("/budget/expense-types");
};

export const CreateExpenseType = async (payload) => {
  return RequestJson("/budget/expense-types", {
    method: "POST",
    body: JSON.stringify(payload)
  });
};

export const UpdateExpenseType = async (typeId, payload) => {
  return RequestJson(`/budget/expense-types/${typeId}`, {
    method: "PUT",
    body: JSON.stringify(payload)
  });
};

export const DeleteExpenseType = async (typeId) => {
  const response = await RequestWithAuth(`/budget/expense-types/${typeId}`, { method: "DELETE" });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(detail || "Request failed");
  }
};

File: apps/ios/EverdayIOS/EverdayIOS/Services/BudgetApi.swift
import Foundation

enum BudgetApi {
    static func fetchIncomeStreams() async throws -> [IncomeStream] {
        try await ApiClient.shared.request(path: "budget/income-streams", requiresAuth: true)
    }

    static func createIncomeStream(_ payload: IncomeStreamCreate) async throws -> IncomeStream {
        try await ApiClient.shared.request(
            path: "budget/income-streams",
            method: "POST",
            body: payload,
            requiresAuth: true
        )
    }

    static func updateIncomeStream(id: Int, payload: IncomeStreamUpdate) async throws -> IncomeStream {
        try await ApiClient.shared.request(
            path: "budget/income-streams/\(id)",
            method: "PUT",
            body: payload,
            requiresAuth: true
        )
    }

    static func deleteIncomeStream(id: Int) async throws {
        try await ApiClient.shared.requestVoid(
            path: "budget/income-streams/\(id)",
            method: "DELETE",
            requiresAuth: true
        )
    }

    static func fetchExpenses() async throws -> [Expense] {
        try await ApiClient.shared.request(path: "budget/expenses", requiresAuth: true)
    }

    static func createExpense(_ payload: ExpenseCreate) async throws -> Expense {
        try await ApiClient.shared.request(
            path: "budget/expenses",
            method: "POST",
            body: payload,
            requiresAuth: true
        )
    }

    static func updateExpense(id: Int, payload: ExpenseUpdate) async throws -> Expense {
        try await ApiClient.shared.request(
            path: "budget/expenses/\(id)",
            method: "PUT",
            body: payload,
            requiresAuth: true
        )
    }

    static func deleteExpense(id: Int) async throws {
        try await ApiClient.shared.requestVoid(
            path: "budget/expenses/\(id)",
            method: "DELETE",
            requiresAuth: true
        )
    }

    static func fetchExpenseAccounts() async throws -> [ExpenseAccount] {
        try await ApiClient.shared.request(path: "budget/expense-accounts", requiresAuth: true)
    }

    static func createExpenseAccount(_ payload: ExpenseAccountCreate) async throws -> ExpenseAccount {
        try await ApiClient.shared.request(
            path: "budget/expense-accounts",
            method: "POST",
            body: payload,
            requiresAuth: true
        )
    }

    static func updateExpenseAccount(id: Int, payload: ExpenseAccountUpdate) async throws -> ExpenseAccount {
        try await ApiClient.shared.request(
            path: "budget/expense-accounts/\(id)",
            method: "PUT",
            body: payload,
            requiresAuth: true
        )
    }

    static func deleteExpenseAccount(id: Int) async throws {
        try await ApiClient.shared.requestVoid(
            path: "budget/expense-accounts/\(id)",
            method: "DELETE",
            requiresAuth: true
        )
    }

    static func fetchExpenseTypes() async throws -> [ExpenseType] {
        try await ApiClient.shared.request(path: "budget/expense-types", requiresAuth: true)
    }

    static func createExpenseType(_ payload: ExpenseTypeCreate) async throws -> ExpenseType {
        try await ApiClient.shared.request(
            path: "budget/expense-types",
            method: "POST",
            body: payload,
            requiresAuth: true
        )
    }

    static func updateExpenseType(id: Int, payload: ExpenseTypeUpdate) async throws -> ExpenseType {
        try await ApiClient.shared.request(
            path: "budget/expense-types/\(id)",
            method: "PUT",
            body: payload,
            requiresAuth: true
        )
    }

    static func deleteExpenseType(id: Int) async throws {
        try await ApiClient.shared.requestVoid(
            path: "budget/expense-types/\(id)",
            method: "DELETE",
            requiresAuth: true
        )
    }

    static func fetchAllocationAccounts() async throws -> [AllocationAccount] {
        try await ApiClient.shared.request(path: "budget/allocation-accounts", requiresAuth: true)
    }

    static func createAllocationAccount(_ payload: AllocationAccountCreate) async throws -> AllocationAccount {
        try await ApiClient.shared.request(
            path: "budget/allocation-accounts",
            method: "POST",
            body: payload,
            requiresAuth: true
        )
    }

    static func updateAllocationAccount(id: Int, payload: AllocationAccountUpdate) async throws -> AllocationAccount {
        try await ApiClient.shared.request(
            path: "budget/allocation-accounts/\(id)",
            method: "PUT",
            body: payload,
            requiresAuth: true
        )
    }

    static func deleteAllocationAccount(id: Int) async throws {
        try await ApiClient.shared.requestVoid(
            path: "budget/allocation-accounts/\(id)",
            method: "DELETE",
            requiresAuth: true
        )
    }
}
